src/v/cluster/tests/leader_balancer_test.cc

```diff
@@ -10,55 +10,354 @@
  */
 #define BOOST_TEST_MODULE leader_balancer
 
+#include "absl/container/flat_hash_map.h"
 #include "cluster/scheduling/leader_balancer_greedy.h"
+#include "leader_balancer_test_utils.h"
+#include "model/metadata.h"
 
+#include <absl/container/flat_hash_set.h>
 #include <boost/test/unit_test.hpp>
 
+#include <algorithm>
+#include <iterator>
+#include <tuple>
 #include <vector>
 
-BOOST_AUTO_TEST_CASE(greedy) {
-    cluster::leader_balancer_strategy::index_type index;
+using index_type = cluster::leader_balancer_strategy::index_type;
+using gbs = cluster::greedy_balanced_shards;
+using reassignment = cluster::leader_balancer_strategy::reassignment;
 
-    // 10 nodes, 2 cores -> 20 shards
-    std::vector<model::broker_shard> shards;
-    for (auto n = 0U; n < 10; n++) {
-        for (auto s = 0U; s < 2; s++) {
-            model::broker_shard shard{model::node_id(n), s};
-            shards.push_back(shard);
-        }
-    }
+/**
+ * @brief Basic validity checks on a reassignment.
+ */
+static void check_valid(const index_type& index, const reassignment& movement) {
+    // from == to is not valid (error should not decrease)
+    BOOST_REQUIRE(movement.from != movement.to);
 
-    size_t replica = 0;
-    for (auto shard : shards) {
-        // 10 groups per shard
-        for (auto g = 0U; g < 10; g++) {
-            raft::group_id group(g);
-            // 3 replica per group
-            std::vector<model::broker_shard> replicas;
-            replicas.push_back(shard); // the "leader"
-            while (replicas.size() != 3) {
-                if (shards[replica % shards.size()] != shard) {
-                    replicas.push_back(shards[replica % shards.size()]);
-                }
-                ++replica;
-            }
-            index[shard][group] = std::move(replicas);
-        }
-    }
+    auto& from_groups = index.at(movement.from);
+    // check that the from shard was the leader
+    BOOST_REQUIRE(from_groups.contains(movement.group));
+
+    // check that the to shard is in the replica set
+    auto& replicas = from_groups.at(movement.group);
+    BOOST_REQUIRE_EQUAL(
+      std::count(replicas.begin(), replicas.end(), movement.to), 1);
+}
+
+BOOST_AUTO_TEST_CASE(greedy_movement) {
+    // 10 nodes
+    // 2 cores x node
+    // 10 partitions per shard
+    // r=3 (3 replicas)
+    auto index = leader_balancer_test_utils::make_cluster_index(10, 2, 10, 3);
 
     auto greed = cluster::greedy_balanced_shards(index, {});
     BOOST_REQUIRE_EQUAL(greed.error(), 0);
 
-    // new groups on shard 5
-    index[shards[5]][raft::group_id(20)] = index[shards[5]][raft::group_id(3)];
-    index[shards[5]][raft::group_id(21)] = index[shards[5]][raft::group_id(3)];
-    index[shards[5]][raft::group_id(22)] = index[shards[5]][raft::group_id(3)];
+    // new groups on shard {2, 0}
+    auto shard20 = model::broker_shard{model::node_id{2}, 0};
```

Comment:
We don't have direct access to the shard array for `shard[5]` anymore, but shard `{2, 0}` should be the same one (though it doesn't really matter: we can pick any arbitrary shard as long as we are consistent).